transports/webrtc/src/upgrade.rs

```diff
@@ -0,0 +1,168 @@
+use libp2p_noise::{Keypair, NoiseConfig, NoiseError, RemoteIdentity, X25519Spec};
+use log::{debug, trace};
+use webrtc::api::APIBuilder;
+use webrtc::data_channel::data_channel_init::RTCDataChannelInit;
+use webrtc::dtls_transport::dtls_role::DTLSRole;
+use webrtc::peer_connection::configuration::RTCConfiguration;
+use webrtc::peer_connection::sdp::session_description::RTCSessionDescription;
+use webrtc_data::data_channel::DataChannel as DetachedDataChannel;
+use webrtc_ice::udp_mux::UDPMux;
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use crate::connection::Connection;
+use crate::connection::PollDataChannel;
+use crate::error::Error;
+use crate::sdp;
+use crate::transport;
+
+pub async fn webrtc(
+    udp_mux: Arc<dyn UDPMux + Send + Sync>,
+    config: RTCConfiguration,
+    addr: Multiaddr,
+    id_keys: identity::Keypair,
+) -> Result<(PeerId, Connection), Error> {
+    trace!("upgrading {}", addr);
+
+    let socket_addr = transport::multiaddr_to_socketaddr(&addr)
+        .ok_or_else(|| Error::InvalidMultiaddr(addr.clone()))?;
+    let our_fingerprint = transport::fingerprint_of_first_certificate(&config);
+
+    let mut se = transport::build_setting_engine(udp_mux, &socket_addr, &our_fingerprint);
+    {
+        // Act as a lite ICE (ICE which does not send additional candidates).
+        se.set_lite(true);
+        // Act as a DTLS server (one which waits for a connection).
+        //
+        // NOTE: removing this seems to break DTLS setup (both sides send `ClientHello` messages,
+        // but none end up responding).
+        se.set_answering_dtls_role(DTLSRole::Server)
+            .map_err(Error::WebRTC)?;
+    }
+    let api = APIBuilder::new().with_setting_engine(se).build();
+    let peer_connection = api.new_peer_connection(config).await?;
+
+    // Set the remote description to the predefined SDP.
+    let remote_fingerprint = if let Some(Protocol::XWebRTC(f)) = addr.iter().last() {
+        transport::fingerprint_to_string(&f)
+    } else {
+        debug!("{} is not a WebRTC multiaddr", addr);
+        return Err(Error::InvalidMultiaddr(addr));
+    };
+    let client_session_description = transport::render_description(
+        sdp::CLIENT_SESSION_DESCRIPTION,
+        socket_addr,
+        &remote_fingerprint,
+    );
+    debug!("OFFER: {:?}", client_session_description);
+    let sdp = RTCSessionDescription::offer(client_session_description).unwrap();
+    peer_connection.set_remote_description(sdp).await?;
+
+    let answer = peer_connection.create_answer(None).await?;
+    // Set the local description and start UDP listeners
+    // Note: this will start the gathering of ICE candidates
+    debug!("ANSWER: {:?}", answer.sdp);
+    peer_connection.set_local_description(answer).await?;
+
+    // Create a datachannel with label 'data'.
+    let data_channel = peer_connection
+        .create_data_channel(
+            "data",
+            Some(RTCDataChannelInit {
+                negotiated: Some(true),
```

Comment:
@ckousik note that with the `negotiated: Some(true),` the server does not initiate the first data channel used for the Noise handshake, but instead waits for the remote (browser) to initiate the client, thus it is in-line with the current libp2p WebRTC specification.

Documentation of the `webrtc-rs` `negotiated` field:

``` Rust
    /// negotiated describes if the data channel is created by the local peer or
    /// the remote peer. The default value of false tells the user agent to
    /// announce the channel in-band and instruct the other peer to dispatch a
    /// corresponding DataChannel. If set to true, it is up to the application
    /// to negotiate the channel and create an DataChannel with the same id
    /// at the other peer.
    pub negotiated: Option<bool>,
```

@melekes please correct me if I am wrong.